File: ios/AITutor/Data/Remote/AiProxyApi.swift
import Foundation
import Functions

/// Tool modes accepted by the OpenAI proxy. Mirrors the `toolMode` switch in
/// `backend/supabase/functions/openai/index.ts` — keep in sync when the
/// backend gains new modes.
enum ToolMode: String {
    case chat
    case interview
    case grammar
    case tutor
    case translate
    case level
    case essayEvaluate = "essay_evaluate"
    case readingGenerate = "reading_generate"
    case listeningGenerate = "listening_generate"
    case taskResponseEvaluate = "task_response_evaluate"
    case speedReadingGenerate = "speed_reading_generate"
    case conversationGenerate = "conversation_generate"
}

struct AiProxyRequest: Encodable {
    let toolMode: String
    let userText: String
    let userLevel: UserLevelDto?
    let examMode: String?

    init(
        toolMode: String,
        userText: String,
        userLevel: UserLevelDto? = nil,
        examMode: String? = nil
    ) {
        self.toolMode = toolMode
        self.userText = userText
        self.userLevel = userLevel
        self.examMode = examMode
    }
}

struct UserLevelDto: Encodable {
    let level: String
    let description: String
}

struct AiProxyResponse: Decodable {
    let text: String
}

struct RealtimeTokenRequest: Encodable {
    let mode: String
}

struct RealtimeTokenResponse: Decodable {
    let token: String
    let sessionId: String
    let expiresAt: Int?
}

/// Thin wrapper around the Supabase Edge Functions that proxy OpenAI calls.
///
/// Edge function names (deployed from `backend/supabase/functions/`):
/// - `openai`          — chat completions (all toolModes)
/// - `realtime-token`  — ephemeral token for OpenAI Realtime (Speaking)
/// - `transcribe`      — Whisper transcription
final class AiProxyApi {
    private let provider: SupabaseProvider

    init(provider: SupabaseProvider = .shared) {
        self.provider = provider
    }

    func callOpenAi(_ request: AiProxyRequest) async throws -> String {
        let response: AiProxyResponse = try await provider.client.functions.invoke(
            "openai",
            options: FunctionInvokeOptions(body: request)
        )
        return response.text
    }

    func createRealtimeToken(mode: String) async throws -> RealtimeTokenResponse {
        let response: RealtimeTokenResponse = try await provider.client.functions.invoke(
            "realtime-token",
            options: FunctionInvokeOptions(body: RealtimeTokenRequest(mode: mode))
        )
        return response
    }

    func transcribeAudio(audioBase64: String) async throws -> String {
        let response: AiProxyResponse = try await provider.client.functions.invoke(
            "transcribe",
            options: FunctionInvokeOptions(body: ["audio": audioBase64])
        )
        return response.text
    }
}

File: android/app/src/main/java/com/aitutor/app/data/remote/AiProxyApi.kt
package com.aitutor.app.data.remote

import io.github.jan.supabase.functions.Functions
import io.github.jan.supabase.functions.functions
import io.ktor.client.call.body
import io.ktor.client.statement.bodyAsText
import io.ktor.http.HttpMethod
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Tool modes accepted by the OpenAI proxy. Mirrors the `toolMode` switch in
 * `_reference/netlify/functions/openai.js` — keep in sync when the backend
 * gains new modes.
 */
enum class ToolMode(val wire: String) {
    CHAT("chat"),
    INTERVIEW("interview"),
    GRAMMAR("grammar"),
    TUTOR("tutor"),
    TRANSLATE("translate"),
    LEVEL("level"),
    ESSAY_EVALUATE("essay_evaluate"),
    READING_GENERATE("reading_generate"),
    LISTENING_GENERATE("listening_generate"),
    TASK_RESPONSE_EVALUATE("task_response_evaluate"),
    SPEED_READING_GENERATE("speed_reading_generate"),
    CONVERSATION_GENERATE("conversation_generate")
}

@Serializable
data class AiProxyRequest(
    @SerialName("toolMode") val toolMode: String,
    @SerialName("userText") val userText: String,
    @SerialName("userLevel") val userLevel: UserLevel? = null,
    @SerialName("examMode") val examMode: String? = null
)

@Serializable
data class UserLevel(
    val level: String,
    val description: String
)

@Serializable
data class AiProxyResponse(
    val text: String
)

@Serializable
data class RealtimeTokenRequest(
    val mode: String
)

@Serializable
data class RealtimeTokenResponse(
    val token: String,
    val sessionId: String,
    val expiresAt: Long? = null
)

/**
 * Thin wrapper around the Supabase Edge Functions that proxy OpenAI calls.
 *
 * Edge function names (deployed from `backend/supabase/functions/`):
 * - `openai`          — chat completions (all toolModes)
 * - `realtime-token`  — ephemeral token for OpenAI Realtime (Speaking)
 * - `transcribe`      — Whisper transcription
 */
@Singleton
class AiProxyApi @Inject constructor(
    private val supabase: SupabaseProvider
) {
    private val functions: Functions get() = supabase.client.functions
    private val json = Json { ignoreUnknownKeys = true; encodeDefaults = false }

    suspend fun callOpenAi(request: AiProxyRequest): String {
        val res = functions.invoke(
            function = "openai",
            body = json.encodeToString(request)
        ) {
            method = HttpMethod.Post
        }
        val raw = res.bodyAsText()
        return json.decodeFromString<AiProxyResponse>(raw).text
    }

    suspend fun createRealtimeToken(mode: String): RealtimeTokenResponse {
        val res = functions.invoke(
            function = "realtime-token",
            body = json.encodeToString(RealtimeTokenRequest(mode))
        ) {
            method = HttpMethod.Post
        }
        return json.decodeFromString(res.bodyAsText())
    }

    suspend fun transcribeAudio(audioBase64: String): String {
        val payload = json.encodeToString(mapOf("audio" to audioBase64))
        val res = functions.invoke(
            function = "transcribe",
            body = payload
        ) {
            method = HttpMethod.Post
        }
        return json.decodeFromString<AiProxyResponse>(res.bodyAsText()).text
    }
}
